packages/itwin/tree-widget/src/tree-widget-react/components/trees/models-tree/internal/ModelsTreeIdsCache.ts

```diff
@@ -376,74 +355,194 @@ export class ModelsTreeIdsCache {
     return entry;
   }
 
-  private async queryCategoryElementCounts(
-    input: Array<{ modelId: Id64String; categoryId: Id64String }>,
-  ): Promise<Array<{ modelId: number; categoryId: number; elementsCount: number }>> {
+  private async queryCategoryElementCounts(input: Array<{ modelId: ModelId; categoryId: CategoryId }>): Promise<Map<ModelCategoryKey, number>> {
+    const result = new Map<ModelCategoryKey, number>();
+    if (input.length === 0) {
+      return result;
+    }
+
+    const reader = this._queryExecutor.createQueryReader(
+      {
+        ecsql: `
+          SELECT COUNT(*) elementsCount, e.Category.Id categoryId, e.Model.Id modelId
+          FROM ${this._hierarchyConfig.elementClassSpecification} e
+          WHERE
+            e.Parent.Id IS NULL AND (
+              ${input.map(({ modelId, categoryId }) => `e.Model.Id = ${modelId} AND e.Category.Id = ${categoryId}`).join(" OR ")}
+            )
+          GROUP BY e.Model.Id, e.Category.Id
+        `,
+      },
+      { rowFormat: "ECSqlPropertyNames", limit: "unbounded" },
+    );
+
+    for await (const row of reader) {
+      const key: ModelCategoryKey = `${row.modelId}-${row.categoryId}`;
+      result.set(key, row.elementsCount);
+    }
+    return result;
+  }
+
+  public async getRootCategoryElementsCount(modelId: ModelId, categoryId: CategoryId): Promise<number> {
+    return this._categoryElementCounts.getRootCategoryElementsCount(modelId, categoryId);
+  }
+
+  private async queryElementsChildrenInfo({ modelId }: { modelId: ModelId }): Promise<ModelParentMap> {
     const reader = this._queryExecutor.createQueryReader(
       {
         ctes: [
-          /* sql */ `
-            CategoryElements(id, modelId, categoryId) AS (
-              SELECT ECInstanceId, Model.Id, Category.Id
+          `
+          ElementsChildrenInfo(Id, CategoryId, ParentId, RootElementCategoryId, ParentHasParent) AS (
```

Comment:
This CTE selects all elements of given model - that could be millions of elements. Did you check how this query performs in large datasets?